runtime/lua/vim/lsp/inlay_hint.lua

```diff
@@ -0,0 +1,262 @@
+--- </pre>
+---
+function M.refresh(opts)
+  opts = opts or {}
+  local bufnr = opts.bufnr
+  bufnr = resolve_bufnr(bufnr or 0)
+  local buffer_windows = {}
+  for _, winid in ipairs(api.nvim_list_wins()) do
+    if api.nvim_win_get_buf(winid) == bufnr then
+      table.insert(buffer_windows, winid)
+    end
+  end
+  for _, window in ipairs(buffer_windows) do
+    local first = vim.fn.line('w0', window)
+    local last = vim.fn.line('w$', window)
+    local params = {
+      textDocument = util.make_text_document_params(bufnr),
+      range = {
+        start = { line = first - 1, character = 0 },
+        ['end'] = { line = last, character = 0 },
+      },
+    }
+    vim.lsp.buf_request(bufnr, 'textDocument/inlayHint', params)
+  end
+  local params = {
+    textDocument = util.make_text_document_params(bufnr),
+    range = {
+      start = { line = 0, character = 0 },
+      ['end'] = { line = api.nvim_buf_line_count(bufnr), character = 0 },
+    },
+  }
+  vim.lsp.buf_request(bufnr, 'textDocument/inlayHint', params)
+end
+
+--- Clear inlay hints
+---
+---@param client_id integer|nil filter by client_id. All clients if nil
+---@param bufnr integer|nil filter by buffer. All buffers if nil
+function M.clear(client_id, bufnr)
+  local buffers = bufnr and { resolve_bufnr(bufnr) } or vim.tbl_keys(hint_cache_by_buf)
+  for _, iter_bufnr in ipairs(buffers) do
+    local bufstate = hint_cache_by_buf[iter_bufnr]
+    local client_lens = (bufstate or {}).client_hint or {}
+    local client_ids = client_id and { client_id } or vim.tbl_keys(client_lens)
+    for _, iter_client_id in ipairs(client_ids) do
+      if bufstate then
+        bufstate.client_hint[iter_client_id] = {}
+      end
+    end
+  end
+  api.nvim__buf_redraw_range(bufnr, 0, -1)
+end
+
+local namespace = api.nvim_create_namespace('vim_lsp_inlayhint')
+
+api.nvim_set_decoration_provider(namespace, {
+  on_win = function(_, _, bufnr, topline, botline)
+    local bufstate = hint_cache_by_buf[bufnr]
+    if not bufstate then
+      return
+    end
+
+    if bufstate.version ~= util.buf_versions[bufnr] then
+      return
+    end
+    local hints_by_client = bufstate.client_hint
+    api.nvim_buf_clear_namespace(bufnr, namespace, 0, -1)
+
+    for lnum = topline, botline do
+      for _, hints_by_lnum in pairs(hints_by_client) do
+        local line_hints = hints_by_lnum[lnum] or {}
+        for _, hint in pairs(line_hints) do
+          local text = ''
+          if type(hint.label) == 'string' then
+            text = hint.label
+          else
+            for _, part in ipairs(hint.label) do
+              text = text .. part.value
+            end
+          end
+          if hint.paddingLeft then
+            text = ' ' .. text
+          end
+          if hint.paddingRight then
+            text = text .. ' '
+          end
+          api.nvim_buf_set_extmark(bufnr, namespace, lnum, hint.position.character, {
```

Comment:
Will this work with all servers? Other places that use `Position` (from LSP) have to convert the given position into character offsets using the client encoding.

This conversion can be expensive, and is better done in the request handler (it requires having the entire line for each character position that needs conversion).